### PR TITLE
fix #1752; html, texMath improvements; webapp, VectorGraphics updates

### LINKED ISSUE #1752: net TABLE bugs
recording for posterity. will fix.
- `net TABLE` incorrectly displays one line tables: compare
```
i1 : needsPackage "Text"; net TABLE {{"one","line"},{"two","lines"}}

o2 = oneline 
     twolines
```
(though spacing could be improved) and
```
i3 : net TABLE {{"one","line"}}

o3 = one 
     line
```
- `net TABLE` does not treat `TR` options correctly:
```
i4 : net TABLE {TR{"one","line","style"=>"background:red"},{"two","lines"}}

o4 = oneline style => background:red
     twolines                       
```

### LINKED ISSUE #1891: doc examples that shouldn't be captured
`help on` redefines `ker` which messes up other doc pages using `ker`, say `help "modules in Macaulay2"`.
so it shouldn't be captured (but rather, run the old-fashioned way in a separate M2 instance).
will fix.

### PR DESCRIPTION
**Pull request by pzinn**
1. fix #1752 -- code for `net TABLE`, `info TABLE` improved.
2. fix #1891
2. `MatrixDegreeExpression` removed -- absorbed into `MatrixExpression`, which has been rewritten.
Some options are not yet used, though already there:
```
i2 : new MatrixExpression from {{{1,2,3},{4,5,6}},BlockMatrix=>{{1,1},{1,2}}}

o2 = | 1 | 2 3 |
     +---+-----+
     | 4 | 5 6 |

o2 : Expression of class MatrixExpression
```
3. `texMathShort` (which was used internally) removed -- replaced with a new function `short` that acts on expressions, visible lists or strings/nets and shortens them. (used to be called `shorten` but I shortened it). Used inernally by some `texMath`; not exported yet, though it might one day be useful for other purposes.
4. minor `texMath`, `html` fixes.
5. minor `webapp.m2`, `VectorGraphics` updates.

Points 3 and 4 are work in progress, but things work as they are and I think it would be good to already incorporate these changes.

### CODE REVIEW

**Comment by pzinn** [comment]
now that's an interesting test failure:
```
--- /home/runner/work/M2/M2/M2/BUILD/build/Macaulay2/tests/ComputationsBook/programming/test.out.expected.trim	2021-01-24 11:22:27.961385071 +0000
+++ /home/runner/work/M2/M2/M2/BUILD/build/Macaulay2/tests/ComputationsBook/programming/test.out.trim	2021-01-24 11:22:27.953385040 +0000
@@ -329,7 +329,7 @@
 
 i66 : toExternalString vars R
 
-o66 = map(R^1,R^{{-1}, {-1}, {-1}},{{x,y,z}})
+o66 = map(R^1,R^{{-1}, {-1}, {-1}},{{x, y, z}})
 ```
This is related to this in ringmap.m2:
```
--net RingMap := f -> net expression f -- can't do that because of annoying spacing issue
```
I never expressionified `net RingMap` because the output looks slightly different due to lack of spacing in the variables.
But eventually I knew it would come back to cause future headaches.
Basically, at the moment `RingMap` violates the golden rule "either define `expression` (but not output routines such as `net`), or define output routines such as `net` (but not `expression`)", and the result is, small fixes as I did to `describe Matrix` will have unintended consequences. In this case, `toExternalString` goes through `expression` and has the extra spacing.

Maybe it's time to fix this properly, that is, uncomment `net RingMap` above and just fix the examples to add the standardized spacing.
Any disagreement?

**Comment by pzinn** [comment]
is there a simple way to see the list of failed tests with error messages?

**Comment by DanGrayson** [comment]
It's fine to modify the tests when spacing in output changes.

**Comment by pzinn** [comment]
OK will take care of the `RingMap` mess and add to this PR

**Comment by mahrud** [comment]
Is this intentional?
```m2
i1 : R = QQ[x];

i2 : toExternalString(R^{{3}})

o2 = R^{{3}}

i3 : toExternalString(R^{{0}})

o3 = R^1

i4 : S = R[y];

i5 : toExternalString(S^{{3,2}})

o5 = S^{{3, 2}}

i6 : toExternalString(S^{{0,0}})

o6 = S^1
```
It's correct, but I think the explicitness of the previous one was useful.

**Comment by pzinn** [comment]
> Is this intentional?
> 
> ```m2
> i1 : R = QQ[x];
> 
> i2 : toExternalString(R^{{3}})
> 
> o2 = R^{{3}}
> 
> i3 : toExternalString(R^{{0}})
> 
> o3 = R^1
> 
> i4 : S = R[y];
> 
> i5 : toExternalString(S^{{3,2}})
> 
> o5 = S^{{3, 2}}
> 
> i6 : toExternalString(S^{{0,0}})
> 
> o6 = S^1
> ```
> 
> It's correct, but I think the explicitness of the previous one was useful.

this is not related to the present PR. the behavior is unchanged AFAICT.

File: M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected [hunk 1]
```diff
@@ -269,7 +269,7 @@ o64 : Matrix R  <--- R
 o65 = matrix {{x, y, z}}
 i66 : toExternalString vars R
 
-o66 = map(R^{{0}}, R^{{-1}, {-1}, {-1}}, {{x, y, z}})
+o66 = map(R^1,R^{{-1}, {-1}, {-1}},{{x, y, z}})
```
Comment (mahrud) [inline M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected:272]: @pzinn This is why I asked.

Comment (pzinn) [reply M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected:272]: I'm not sure how up to date some of these "expected" files are.

Comment (mahrud) [reply M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected:272]: Oh, I see. they do get checked regularly, but with `test.out.expected`, not `chapter.out.expected`. You should change the former only, I believe.

Comment (pzinn) [reply M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected:272]: As far as I understand the only one that's tested is test.out.expected?
I actually fixed the other one (chapter.out.expected) by mistake.

Comment (pzinn) [reply M2/Macaulay2/tests/ComputationsBook/programming/chapter.out.expected]: OK. undone.

**Comment by DanGrayson** [comment]
@mikestillman -- have you seen this crash before?

```m2
 -- capturing check(74, "MinimalPrimes")                                    -- SIGSEGV
-* stack trace, pid: 135063
 0# stack_trace(std::ostream&, bool) at ../../../../Macaulay2/d/main.cpp:124
 1# segv_handler at ../../../../Macaulay2/d/main.cpp:241
 2# 0x00007F8EAAA28210 in /lib/x86_64-linux-gnu/libc.so.6
 3# FractionField::mult(ring_elem, ring_elem) const at ../../../../Macaulay2/e/frac.cpp:516
 4# GBRingPoly::mult_by_term1(FreeModule const*, gbvector const*, ring_elem, int const*, int) at ../../../../Macaulay2/e/gbring.cpp:116
 5# GBRing::mult_by_term(FreeModule const*, gbvector const*, ring_elem, int const*, int) at ../../../../Macaulay2/e/gbring.cpp:837
 6# GBRing::gbvector_reduce_lead_term(FreeModule const*, FreeModule const*, gbvector*, gbvector*&, gbvector*&, gbvector const*, gbvector const*, bool, ring_elem&) at ../../../../Macaulay2/e/gbring.cpp:967
 7# ReducedGB_Field::remainder(POLY&, bool, ring_elem&) at ../../../../Macaulay2/e/reducedgb-field.cpp:124
 8# ReducedGB::matrix_lift(Matrix const*, Matrix const**, Matrix const**) at ../../../../Macaulay2/e/reducedgb.cpp:191
 9# PolyRingQuotient::ann(ring_elem, ring_elem) const at ../../../../Macaulay2/e/polyquotient.cpp:324
10# PolyRingQuotient::invert(ring_elem) const at ../../../../Macaulay2/e/polyquotient.cpp:186
11# Ring::power(ring_elem, __mpz_struct const*) const at ../../../../Macaulay2/e/ring.cpp:112
12# RingElement::power(__mpz_struct const*) const at ../../../../Macaulay2/e/relem.cpp:125
13# engine_circ__1 at ../../../../Macaulay2/d/engine.dd:140
14# actors_circ_ at ../../../../Macaulay2/d/actors.d:614
```

**Comment by d-torrance** [comment]
> @mikestillman -- have you seen this crash before?
> 
> ```m2
>  -- capturing check(74, "MinimalPrimes")                                    -- SIGSEGV
> -* stack trace, pid: 135063
>  0# stack_trace(std::ostream&, bool) at ../../../../Macaulay2/d/main.cpp:124
>  1# segv_handler at ../../../../Macaulay2/d/main.cpp:241
>  2# 0x00007F8EAAA28210 in /lib/x86_64-linux-gnu/libc.so.6
>  3# FractionField::mult(ring_elem, ring_elem) const at ../../../../Macaulay2/e/frac.cpp:516
>  4# GBRingPoly::mult_by_term1(FreeModule const*, gbvector const*, ring_elem, int const*, int) at ../../../../Macaulay2/e/gbring.cpp:116
>  5# GBRing::mult_by_term(FreeModule const*, gbvector const*, ring_elem, int const*, int) at ../../../../Macaulay2/e/gbring.cpp:837
>  6# GBRing::gbvector_reduce_lead_term(FreeModule const*, FreeModule const*, gbvector*, gbvector*&, gbvector*&, gbvector const*, gbvector const*, bool, ring_elem&) at ../../../../Macaulay2/e/gbring.cpp:967
>  7# ReducedGB_Field::remainder(POLY&, bool, ring_elem&) at ../../../../Macaulay2/e/reducedgb-field.cpp:124
>  8# ReducedGB::matrix_lift(Matrix const*, Matrix const**, Matrix const**) at ../../../../Macaulay2/e/reducedgb.cpp:191
>  9# PolyRingQuotient::ann(ring_elem, ring_elem) const at ../../../../Macaulay2/e/polyquotient.cpp:324
> 10# PolyRingQuotient::invert(ring_elem) const at ../../../../Macaulay2/e/polyquotient.cpp:186
> 11# Ring::power(ring_elem, __mpz_struct const*) const at ../../../../Macaulay2/e/ring.cpp:112
> 12# RingElement::power(__mpz_struct const*) const at ../../../../Macaulay2/e/relem.cpp:125
> 13# engine_circ__1 at ../../../../Macaulay2/d/engine.dd:140
> 14# actors_circ_ at ../../../../Macaulay2/d/actors.d:614
> ```

I've seen it -- https://github.com/Macaulay2/M2/issues/1742#issuecomment-751519232

**Comment by DanGrayson** [comment]
A new function called "short" is being exported, but there doesn't seem to be any documentation.  What does it do?

**Comment by pzinn** [comment]
oops. as the list of changes originally said, `short` was not meant to be exported. so let me revert that so it's not exported (yet) since it's only used internally for now by some `texMath`.

**Comment by DanGrayson** [comment]
One of the tests failed because of differences in spacing in the output for the test in `Macaulay2/tests/ComputationsBook/completeIntersections`.  That will have to be fixed.

**Comment by pzinn** [comment]
OK will do. @mahrud I'm not able to run `ctest` on the ComputationBooks examples, keeps giving me error messages like
```
3372: /usr/bin/diff: /home/pzinn/M2/M2/BUILD/fedora/Macaulay2/tests/ComputationsBook/toricHilbertScheme/test.out.expected.trim: No such file or directory
3372: /usr/bin/diff: /home/pzinn/M2/M2/BUILD/fedora/Macaulay2/tests/ComputationsBook/toricHilbertScheme/test.out.trim: No such file or directory

```
so I've been testing things manually and must have missed some.

**Comment by mahrud** [comment]
Run `ninja M2-tests` first.

**Comment by pzinn** [comment]
thanks

**Pull request merged**